
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Quick start guide &#8212; xsmc 1.0.3.dev0+gfd916e6.d20201021 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Citing xsmc" href="cite.html" />
    <link rel="prev" title="Installation instructions" href="install.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cite.html" title="Citing xsmc"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation instructions"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">xsmc 1.0.3.dev0+gfd916e6.d20201021 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Quick start guide</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Quick-start-guide">
<h1>Quick start guide<a class="headerlink" href="#Quick-start-guide" title="Permalink to this headline">¶</a></h1>
<p>This packages implements methods described in <a class="reference external" href="https://doi.org/10.1101/2020.09.21.307355">the paper</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Exact</span> <span class="n">decoding</span> <span class="n">of</span> <span class="n">the</span> <span class="n">sequentially</span> <span class="n">Markov</span> <span class="n">coalescent</span>
<span class="n">Caleb</span> <span class="n">Ki</span><span class="p">,</span> <span class="n">Jonathan</span> <span class="n">Terhorst</span>
<span class="n">bioRxiv</span> <span class="mf">2020.09</span><span class="o">.</span><span class="mf">21.307355</span><span class="p">;</span> <span class="n">doi</span><span class="p">:</span><span class="mf">10.1101</span><span class="o">/</span><span class="mf">2020.09</span><span class="o">.</span><span class="mf">21.307355</span>
</pre></div>
</div>
<p>The methods are implemented in a Python package called <code class="docutils literal notranslate"><span class="pre">xsmc</span></code>. The package has two main uses: drawing exact samples from the posterior distribution, and fast estimation of the maximum <em>a posteriori</em> (MAP) path. Convenience methods are also included for plotting.</p>
<p>We will demonstrate how the method works on simulated data using <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">simulate</span><span class="p">(</span><span class="n">recombination_rate</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">sample_size</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">msp</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
        <span class="n">sample_size</span><span class="o">=</span><span class="n">sample_size</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
        <span class="n">Ne</span><span class="o">=</span><span class="mf">1e4</span><span class="p">,</span>
        <span class="n">length</span><span class="o">=</span><span class="n">length</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">4e9</span><span class="p">)),</span>
        <span class="n">demographic_events</span><span class="o">=</span><span class="p">[</span>
            <span class="n">msp</span><span class="o">.</span><span class="n">PopulationParametersChange</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mf">1e5</span><span class="p">),</span>
            <span class="n">msp</span><span class="o">.</span><span class="n">PopulationParametersChange</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mf">1000.0</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mf">1e4</span><span class="p">),</span>
        <span class="p">],</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 3;
                var nbb_unformatted_code = "def simulate(recombination_rate, length, sample_size=2):\n    return msp.simulate(\n        sample_size=sample_size,\n        mutation_rate=1e-8,\n        recombination_rate=recombination_rate,\n        Ne=1e4,\n        length=length,\n        random_seed=np.random.randint(1, int(4e9)),\n        demographic_events=[\n            msp.PopulationParametersChange(time=0.0, initial_size=1e5),\n            msp.PopulationParametersChange(time=1000.0, initial_size=1e4),\n        ],\n    )";
                var nbb_formatted_code = "def simulate(recombination_rate, length, sample_size=2):\n    return msp.simulate(\n        sample_size=sample_size,\n        mutation_rate=1e-8,\n        recombination_rate=recombination_rate,\n        Ne=1e4,\n        length=length,\n        random_seed=np.random.randint(1, int(4e9)),\n        demographic_events=[\n            msp.PopulationParametersChange(time=0.0, initial_size=1e5),\n            msp.PopulationParametersChange(time=1000.0, initial_size=1e4),\n        ],\n    )";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>All methods are accessed by creating an instance of the <code class="docutils literal notranslate"><span class="pre">xsmc.XSMC</span></code> class. The class has three required arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ts</span></code>, a <code class="docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code> instance containing the data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">focal</span></code>, the node id of the focal haplotype in <code class="docutils literal notranslate"><span class="pre">ts</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">panel</span></code>, a list of node ids in <code class="docutils literal notranslate"><span class="pre">ts</span></code> which are potential targets for <code class="docutils literal notranslate"><span class="pre">focal</span></code> to copy from.</p></li>
</ul>
<p>First, we focus on the simplest type of analysis, where we have two chromosomes sampled from a singled diploid individual.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">xsmc</span> <span class="kn">import</span> <span class="n">XSMC</span><span class="p">,</span> <span class="n">Segmentation</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-9</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mf">1e6</span><span class="p">,</span> <span class="n">sample_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">xs</span> <span class="o">=</span> <span class="n">XSMC</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">focal</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">panel</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">rho_over_theta</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 4;
                var nbb_unformatted_code = "from xsmc import XSMC, Segmentation\n\ndata = simulate(recombination_rate=1e-9, length=1e6, sample_size=2)\nxs = XSMC(data, focal=0, panel=[1], rho_over_theta=0.1)";
                var nbb_formatted_code = "from xsmc import XSMC, Segmentation\n\ndata = simulate(recombination_rate=1e-9, length=1e6, sample_size=2)\nxs = XSMC(data, focal=0, panel=[1], rho_over_theta=0.1)";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<div class="section" id="Exact-sampling">
<h2>Exact sampling<a class="headerlink" href="#Exact-sampling" title="Permalink to this headline">¶</a></h2>
<p>To sample paths from the posterior, use the <code class="docutils literal notranslate"><span class="pre">XSMC.sample()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">paths</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 5;
                var nbb_unformatted_code = "paths = xs.sample(k=100, seed=1)";
                var nbb_formatted_code = "paths = xs.sample(k=100, seed=1)";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>The returned object is a list of length <code class="docutils literal notranslate"><span class="pre">k</span></code> of <code class="docutils literal notranslate"><span class="pre">Segmentation</span></code> objects. These objects can be easily plotted to visualize the posterior:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_10_0.png" src="_images/quickstart_10_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 6;
                var nbb_unformatted_code = "for p in paths:\n    p.draw(alpha=0.1)\nplt.yscale(\"log\")";
                var nbb_formatted_code = "for p in paths:\n    p.draw(alpha=0.1)\nplt.yscale(\"log\")";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>To compare the posterior with the truth, we can use built-in convenience methods to turn data into a <code class="docutils literal notranslate"><span class="pre">Segmentation</span></code> object and plot it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">truth</span> <span class="o">=</span> <span class="n">Segmentation</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">focal</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">panel</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">truth</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_12_0.png" src="_images/quickstart_12_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 7;
                var nbb_unformatted_code = "truth = Segmentation.from_ts(data, focal=0, panel=[1])\ntruth.draw()\nplt.yscale(\"log\")";
                var nbb_formatted_code = "truth = Segmentation.from_ts(data, focal=0, panel=[1])\ntruth.draw()\nplt.yscale(\"log\")";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<div class="section" id="Rescaling-the-output">
<h3>Rescaling the output<a class="headerlink" href="#Rescaling-the-output" title="Permalink to this headline">¶</a></h3>
<p>You may notice that the <span class="math notranslate nohighlight">\(y\)</span> scale of the above plots is different. By default, <code class="docutils literal notranslate"><span class="pre">Segmentation</span></code>s returned by <code class="docutils literal notranslate"><span class="pre">XSMC</span></code> are in units of coalescent time. This matches the scaling of the Li &amp; Durbin’s PSMC.</p>
<p>To convert these times to generations, you must know the biological mutation rate <span class="math notranslate nohighlight">\(\mu\)</span>, and then estimate <span class="math notranslate nohighlight">\(N_0\)</span> by the formula <span class="math notranslate nohighlight">\(N_0 = \theta / (4 \mu)\)</span>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">N0</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">theta</span> <span class="o">/</span> <span class="p">(</span><span class="mi">4</span> <span class="o">*</span> <span class="mf">1e-8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 8;
                var nbb_unformatted_code = "N0 = xs.theta / (4 * 1e-8)";
                var nbb_formatted_code = "N0 = xs.theta / (4 * 1e-8)";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>Then, you can convert time to generations by multiplying by <span class="math notranslate nohighlight">\(2N_0\)</span>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">rescale</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N0</span><span class="p">)</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;tab:blue&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span>
    <span class="p">)</span>  <span class="c1"># divide by the mutation rate used in the simulation, convert to haploid</span>
<span class="n">truth</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;tab:red&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="c1"># plt.ylim(0, 1e4)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_16_0.png" src="_images/quickstart_16_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 9;
                var nbb_unformatted_code = "for p in paths:\n    p.rescale(2 * N0).draw(\n        color=\"tab:blue\", alpha=0.1\n    )  # divide by the mutation rate used in the simulation, convert to haploid\ntruth.draw(color=\"tab:red\")\nplt.yscale(\"log\")\n# plt.ylim(0, 1e4)";
                var nbb_formatted_code = "for p in paths:\n    p.rescale(2 * N0).draw(\n        color=\"tab:blue\", alpha=0.1\n    )  # divide by the mutation rate used in the simulation, convert to haploid\ntruth.draw(color=\"tab:red\")\nplt.yscale(\"log\")\n# plt.ylim(0, 1e4)";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
</div>
<div class="section" id="Computing-functionals-of-the-posterior">
<h3>Computing functionals of the posterior<a class="headerlink" href="#Computing-functionals-of-the-posterior" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">Segmentation.to_pp()</span></code> method can be used to get access to the raw stepwise function it represents (returned as a <strong>p</strong>iecewise <strong>p</strong>olynomial, whence the name). By calling this method on paths returned from <code class="docutils literal notranslate"><span class="pre">XSMC.sample_paths()</span></code>, this you can compute functionals of the posterior distribution.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">get_sequence_length</span><span class="p">(),</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">p</span><span class="o">.</span><span class="n">to_pp</span><span class="p">()(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">])</span>
<span class="n">q</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">N0</span>
<span class="p">)</span>  <span class="c1"># faster than calling .rescale() on each path</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">q</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">q</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">q</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">truth</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_18_0.png" src="_images/quickstart_18_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 10;
                var nbb_unformatted_code = "x = np.linspace(0, data.get_sequence_length(), 1000)\ny = np.array([p.to_pp()(x) for p in paths])\nq = (\n    np.quantile(y, [0.25, 0.5, 0.75], axis=0) * 2 * N0\n)  # faster than calling .rescale() on each path\nplt.plot(x, q[1])\nplt.fill_between(x, q[0], q[2], alpha=0.3)\ntruth.draw()\nplt.yscale(\"log\")";
                var nbb_formatted_code = "x = np.linspace(0, data.get_sequence_length(), 1000)\ny = np.array([p.to_pp()(x) for p in paths])\nq = (\n    np.quantile(y, [0.25, 0.5, 0.75], axis=0) * 2 * N0\n)  # faster than calling .rescale() on each path\nplt.plot(x, q[1])\nplt.fill_between(x, q[0], q[2], alpha=0.3)\ntruth.draw()\nplt.yscale(\"log\")";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
</div>
<div class="section" id="Multiple-samples">
<h3>Multiple samples<a class="headerlink" href="#Multiple-samples" title="Permalink to this headline">¶</a></h3>
<p>We can analyze multiple samples using a “trunk approximation” described in the paper. In this case, the returned segmentation approximates the earliest time to coalescence with any member of the panel.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mf">1e6</span><span class="p">,</span> <span class="n">sample_size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">focal</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">panel</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="n">xs</span> <span class="o">=</span> <span class="n">XSMC</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">focal</span><span class="o">=</span><span class="n">focal</span><span class="p">,</span> <span class="n">panel</span><span class="o">=</span><span class="n">panel</span><span class="p">,</span> <span class="n">rho_over_theta</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="n">paths</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">k</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 11;
                var nbb_unformatted_code = "data = simulate(recombination_rate=1e-8, length=1e6, sample_size=20)\nfocal = 0\npanel = list(range(1, 20))\nxs = XSMC(data, focal=focal, panel=panel, rho_over_theta=1.0)\npaths = xs.sample(k=100, seed=1)";
                var nbb_formatted_code = "data = simulate(recombination_rate=1e-8, length=1e6, sample_size=20)\nfocal = 0\npanel = list(range(1, 20))\nxs = XSMC(data, focal=focal, panel=panel, rho_over_theta=1.0)\npaths = xs.sample(k=100, seed=1)";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">N0</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">theta</span> <span class="o">/</span> <span class="p">(</span><span class="mi">4</span> <span class="o">*</span> <span class="mf">1e-8</span><span class="p">)</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">rescale</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N0</span><span class="p">)</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;tab:blue&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<span class="n">truth</span> <span class="o">=</span> <span class="n">Segmentation</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">focal</span><span class="o">=</span><span class="n">focal</span><span class="p">,</span> <span class="n">panel</span><span class="o">=</span><span class="n">panel</span><span class="p">)</span>
<span class="n">truth</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;tab:red&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_21_0.png" src="_images/quickstart_21_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 12;
                var nbb_unformatted_code = "N0 = xs.theta / (4 * 1e-8)\nfor p in paths:\n    p.rescale(2 * N0).draw(color=\"tab:blue\", alpha=0.05)\ntruth = Segmentation.from_ts(data, focal=focal, panel=panel)\ntruth.draw(color=\"tab:red\")\nplt.yscale(\"log\")";
                var nbb_formatted_code = "N0 = xs.theta / (4 * 1e-8)\nfor p in paths:\n    p.rescale(2 * N0).draw(color=\"tab:blue\", alpha=0.05)\ntruth = Segmentation.from_ts(data, focal=focal, panel=panel)\ntruth.draw(color=\"tab:red\")\nplt.yscale(\"log\")";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
</div>
</div>
<div class="section" id="Viterbi-Decoding">
<h2>Viterbi Decoding<a class="headerlink" href="#Viterbi-Decoding" title="Permalink to this headline">¶</a></h2>
<p>The other function of the <code class="docutils literal notranslate"><span class="pre">XSMC</span></code> class is to compute the Viterbi/most probable path under the posterior:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data</span> <span class="o">=</span> <span class="n">simulate</span><span class="p">(</span><span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-9</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mf">10e6</span><span class="p">,</span> <span class="n">sample_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">xs</span> <span class="o">=</span> <span class="n">XSMC</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">rho_over_theta</span><span class="o">=</span><span class="mf">1e-9</span> <span class="o">/</span> <span class="mf">1.4e-8</span><span class="p">)</span>
<span class="n">N0</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">theta</span> <span class="o">/</span> <span class="p">(</span><span class="mi">4</span> <span class="o">*</span> <span class="mf">1e-8</span><span class="p">)</span>
<span class="n">map_path</span> <span class="o">=</span> <span class="n">xs</span><span class="o">.</span><span class="n">viterbi</span><span class="p">()</span>
<span class="n">map_path</span><span class="o">.</span><span class="n">rescale</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N0</span><span class="p">)</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;viterbi&quot;</span><span class="p">)</span>
<span class="n">Segmentation</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;truth&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x7f0493c47190&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/quickstart_23_1.png" src="_images/quickstart_23_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 13;
                var nbb_unformatted_code = "data = simulate(recombination_rate=1e-9, length=10e6, sample_size=10)\nxs = XSMC(data, 0, [1, 2], rho_over_theta=1e-9 / 1.4e-8)\nN0 = xs.theta / (4 * 1e-8)\nmap_path = xs.viterbi()\nmap_path.rescale(2 * N0).draw(label=\"viterbi\")\nSegmentation.from_ts(data, 0, [1, 2]).draw(label=\"truth\")\nplt.yscale(\"log\")\nplt.legend()";
                var nbb_formatted_code = "data = simulate(recombination_rate=1e-9, length=10e6, sample_size=10)\nxs = XSMC(data, 0, [1, 2], rho_over_theta=1e-9 / 1.4e-8)\nN0 = xs.theta / (4 * 1e-8)\nmap_path = xs.viterbi()\nmap_path.rescale(2 * N0).draw(label=\"viterbi\")\nSegmentation.from_ts(data, 0, [1, 2]).draw(label=\"truth\")\nplt.yscale(\"log\")\nplt.legend()";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>Note that each path also includes a haplotype component which is not plotted:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">map_path</span><span class="o">.</span><span class="n">segments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Segment(hap=2, interval=array([     0, 476205]), height=0.08770837623933274, mutations=21)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="output_javascript"></div>
<script type="text/javascript">
var element = document.currentScript.previousSibling.previousSibling;

            setTimeout(function() {
                var nbb_cell_id = 14;
                var nbb_unformatted_code = "map_path.segments[0]";
                var nbb_formatted_code = "map_path.segments[0]";
                var nbb_cells = Jupyter.notebook.get_cells();
                for (var i = 0; i < nbb_cells.length; ++i) {
                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {
                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {
                             nbb_cells[i].set_text(nbb_formatted_code);
                        }
                        break;
                    }
                }
            }, 500);

</script></div>
</div>
<p>This says that the first segment in the MAP path had height 0.09 (in coalecent units) and coalesced with panel haplotype 2.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quick start guide</a><ul>
<li><a class="reference internal" href="#Exact-sampling">Exact sampling</a><ul>
<li><a class="reference internal" href="#Rescaling-the-output">Rescaling the output</a></li>
<li><a class="reference internal" href="#Computing-functionals-of-the-posterior">Computing functionals of the posterior</a></li>
<li><a class="reference internal" href="#Multiple-samples">Multiple samples</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Viterbi-Decoding">Viterbi Decoding</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="install.html"
                        title="previous chapter">Installation instructions</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="cite.html"
                        title="next chapter">Citing <cite>xsmc</cite></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/quickstart.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cite.html" title="Citing xsmc"
             >next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation instructions"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">xsmc 1.0.3.dev0+gfd916e6.d20201021 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Quick start guide</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Caleb Ki and Jonathan Terhorst.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.2.1.
    </div>
  </body>
</html>